ibis/expr/datatypes/pyarrow.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+import functools
```

Comment:
@cpcloud could you create a follow-up to make this pluggable from backends?

Reply:
@kszucs I think that's a good idea, but I think you probably have a better idea of what to do!

Would you mind making that issue?

Reply:
Sure, going to create that.